### PR TITLE
Metrics

### PR DESCRIPTION
**Pull request by telegraphroad**

### CODE REVIEW

**Comment by vedran-kasalica** [comment]
Hi @telegraphroad, thank you for making the PR. I see that the tests are failing often with the statement `No module named 'src'`. 

I think that the issue is with the way you import methods, for example:
`from src.faivor.metrics.classification.performance import ClassificationPerformanceMetrics`

You should drop `src`:
`from faivor.metrics.classification.performance import ClassificationPerformanceMetrics`

We define the library as `faivor` [here](https://github.com/MaastrichtU-BISS/FAIVOR-ML-Validator/blob/e0f7819b4b0044b636c580b488eec69343a08a02/pyproject.toml#L8C1-L9C42).

File: src/faivor/metrics/test.py
Comment (vedran-kasalica) [inline src/faivor/metrics/test.py:1]: Do we need this file, it looks like a copy of [tests/faivor/metrics/classification/test_performance.py](https://github.com/MaastrichtU-BISS/FAIVOR-ML-Validator/pull/3/files#diff-635a7ab929423d8b433d80d1e3fb87e5b1444d704a7dffb32300432c9f4406fb)
Viewed
Original file line number	Diff line number	Diff line change
@@ -0,0 +1,32 @@

File: tests/faivor/metrics/classification/test_explainability.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import unittest
```
Comment (vedran-kasalica) [inline tests/faivor/metrics/classification/test_explainability.py:1]: We tend to use `pytest` instead of `unittest` as it provides cleaner code. For example you don't need 

```
if __name__ == '__main__':
    unittest.main()
```
because you would simply run command `pytest .` and it would execute all methods of format `test_` in modules of format `test_*.py`. 

In addition, you don't need to create the class object, but just list the methods, and instead of e.g., `setUp` you can define as many additional methods as you need, as long as they don't start with `test_`.

File: tests/faivor/metrics/classification/test_explainability.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+                        result = getattr(self.metrics, name)(self.probabilities_class)
+                    else:
+                        result = getattr(self.metrics, name)(self.y_true_class, self.y_pred_class)
+                    self.assertIsNotNone(result, f"Metric {name} returned None")
+                except Exception as e:
+                    self.fail(f"Metric {name} raised an exception: {e}")
+
+
+
+if __name__ == '__main__':
```
Comment (vedran-kasalica) [inline tests/faivor/metrics/classification/test_explainability.py:34]: this is not needed when using `pytest`

File: tests/faivor/metrics/classification/test_explainability.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    unittest.main()
```
Comment (vedran-kasalica) [inline tests/faivor/metrics/classification/test_explainability.py:35]:
```suggestion

```
this is not needed when using `pytest`

File: tests/faivor/metrics/classification/test_explainability.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import unittest
+import numpy as np
+import torch
+from src.faivor.metrics.classification.explainability import ClassificationExplainabilityMetrics
+
+
+class TestClassificationExplainabilityMetrics(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (telegraphroad) [reply tests/faivor/metrics/classification/test_explainability.py:8]: Hey @vedran-kasalica , thanks for the feedback and the suggestions! The src thing was working on my machine so I missed the error on actions. I fixed it, also had some module import issues which were finally fixed by adding init files to the test subfolders so each suit of tests would be treated as submodule.

The test.py file was just there to provide a more direct way to iterate through all the metrics. I've removed it for now, can add it as a notebook later if needed.

Also moved the tests to pytest.

[the remark this answers is not in the document]
Comment (vedran-kasalica) [reply tests/faivor/metrics/classification/test_explainability.py:8]: Thank you! Indeed, a notebook would be a good way to demonstrate the library.

I will look into the PR today or tomorrow more thoroughly, the first comments were there to fix the CI/CD.

File: src/faivor/metrics/classification/explainability.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    """Base class for classification explainability metrics."""
+    pass
+
+
+class ClassificationExplainabilityMetrics(BaseClassificationExplainabilityMetrics, metaclass=ClassificationExplainabilityMetricsMeta):
+    """Class for classification explainability metrics."""
+
+    def custom_prediction_entropy(self, probas):
+        """Calculate the average entropy of prediction probabilities."""
+        import numpy as np
```
Comment (vedran-kasalica) [inline src/faivor/metrics/classification/explainability.py:33]: It is recommended to keep all the imports at the top of the document

**Comment by vedran-kasalica** [comment]
I added a commit to remove all `unittests` as they were still used. Feel free to see the changes from the last commit https://github.com/MaastrichtU-BISS/FAIVOR-ML-Validator/pull/3/commits/85320d2c69c6a061054d240150b38ec856e9e01a

**Comment by telegraphroad** [comment]
todo:
- sanity
- check metric documentations
- error handling for metric imports
- metrics in config file
- one liner description for metrics

File: src/faivor/metrics/metric.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
+from dataclasses import dataclass, field
+from typing import Callable, Dict
+from abc import ABC
+import torch
+
+@dataclass
+class ModelMetric(ABC):
```
Comment (vedran-kasalica) [inline src/faivor/metrics/metric.py:7]:
```suggestion
import torch

@dataclass
class ModelMetric():
```
Is there a particular reason to make the class abstract? We are instantiating it later in the code, so I assume that it was not attended to be an abstract class.

**Comment by telegraphroad** [comment]
You're right! The class is instantiated directly, so there's no need for it to be abstract.

**Comment by vedran-kasalica** [comment]
The changes look good, thanks for the updates! I created an [issue](https://github.com/MaastrichtU-BISS/FAIVOR-ML-Validator/issues/5) to ensure we don't forget to test the configured metrics.
